docs/AsyncWrap/README.md

```diff
@@ -1,15 +1,202 @@
+disabled, the handle objects there are already being tracked will continue
+to be tracked. Finally there are some cases where handle objects have a parent.
+For example a TCP server creates TCP sockets, thus the socket handles have the
+server handle as a parent. In this case AsyncWrap should be enabled before
+the server handle is created, for the TCP sockets to be tracked.
+
+Disabling AsyncWrap can be done with:
+
+```javascript
+asyncWrap.disable();
+```
+
+#### The hooks
+
+Currently there are 3 hooks: `init`, `before` and `after`. The function
+signatures are quite similar. In all cases the `this` variable is the handle
+object. Users may read properties from this object such as `port` and `address` in the `TCPConnectWrap` case, or set user specific properties.
+
+Finally the `init` has an extra argument called `provider`, this is an
+integer that refer to names defined in an `async.Providers` object map.
+
+## example
+
+A classic use case for AsyncWrap is to create a long-stack-trace tool.
+
+```javascript
+const asyncWrap = process.binding('async_wrap');
+
+asyncWrap.setupHooks(init, before, after);
+asyncWrap.enable();
+
+// global state variable, that contains the current stack trace
+let currentStack = '';
+
+function init(provider) {
+  // When a handle is created, collect the stack trace such that we later
+  // can see what involved the handle constructor.
+  const localStack = (new Error()).stack.split('\n').slice(1).join('\n');
+  // Compute the full stack and store it as a property on the handle object,˛
+  // such that it can be fetched later in either `before` or `after`.
+  this._full_init_stack = localStack + '\n' + currentStack;
+}
+function before() {
+  // A callback is about to be called, update the `currentStack` such that
+  // it is correct for when another handle is initialized or `getStack` is called.
+  currentStack = this._full_init_stack;
+}
+function after() {
+  // At the time of writing there are some odd cases where there is no handle
+  // context, this line prevents that from resulting in wrong stack trace. But
+  // the stack trace will be shorter compared to what ideally should happen.
+  currentStack = '';
+}
+
+function getStack(message) {
+  const localStack = new Error(message);
+  return localStack.stack + '\n' + currentStack;
+}
+module.exports = getStack;
+```
+
+Please note that this example is way simpler, than what is required from a
+complete long-stack-trace implementation.
+
+## Things you might not expect
+
+* `console.log` is async and thus invokes AsyncWrap, thus using `console.log`
+inside one of the hooks, creates an infinite recursion. Use `fs.syncWrite(1, msg)`
+or `process._rawDebug(msg)` instead.
+
+* `process.nextTick` never creates a handle object. You will have to money patch
+this.
+
+* Timer functions (like `setTimeout`) shares a single Timer handle, thus you
+will usually have to money patch those functions.
```

Comment:
Another thing to add would be that because of some quirks in the node API, disabling at certain times can lead to unexpected outcomes. For example:
```js
async_wrap.enable();
let server = net.createServer();
async_wrap.disable();
server.listen(0);
```
In the above the server will not be tracked. Because the `TCPWrap` is not instantiated until `listen` is called. Reason for this is because we don't know whether a PIPE or TCP will be used until that time. It's been very frustrating for me, but not something I've been able to easily work around.